Reset any branch to selected commit

pmiossec opened the pull request:
Fixes #2876
Implementation initially based on @mdonatas


## Proposed changes

- In the revision gird, being able to reset any branch (not only the one checked out) to the selected commit, using "git reset --force"

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/460196/79908466-91018200-841b-11ea-9d5b-7131346912ec.png)

### After

![image](https://user-images.githubusercontent.com/6248932/80648730-f505f500-8a70-11ea-9e1f-688ec2c5a427.png)

The initial view when opening
![image](https://user-images.githubusercontent.com/6248932/80648803-1c5cc200-8a71-11ea-89ca-16164e5fdf53.png)

The new reset form for "other branches"
![image](https://user-images.githubusercontent.com/6248932/80648855-34344600-8a71-11ea-95c3-1cf55f5f8280.png)

## Test methodology <!-- How did you ensure quality? -->

- Manual

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 3.3.0
- Build 0a69f31ca246ebe13a212db80476180c4ac662d8
- Git 2.25.0.windows.1 (recommended: 2.25.1 or later)
- Microsoft Windows NT 6.2.9200.0
- .NET Framework 4.8.4150.0
- DPI 192dpi (200% scaling)


<!-- Mention language, UI scaling, or anything else that might be relevant -->


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented:
/cc @mstv

gerhardol commented:
I rather would like to see a new menu item (probably in Advanced): Reset a branch to here...
(The label is misleading: Reset Current branch to here..., and resetting the current branch is still more common - at least for me).

mstv commented:
Yes, I want to have two menu items: "Reset current branch to here..." as before and "Reset a branch to here..." right below, not hidden in `Advanced`. One level of submenu is 

`build.cmd -loc` ignores `FormResetBranch`, @RussKie. At least the `GitUITests.TranslationTest` fails.

gerhardol commented:
This PR eliminates the possibility to reset current branch, useful to remove garbage in Index/worktree
Sure, you can still use Commands-Reset, Commit-Reset or RevDiff-select-in-diff-Reset-to-first

Text incorrect in DarkTheme
Add the following line in FormResetBranch() constructor ( I considered pushing to your branch).

```            labelResetBranchWarning.SetForeColorForBackColor();
```
(the overlapping text is fixed in master for this PR)
before:
![image](https://user-images.githubusercontent.com/6248932/79987449-b8a12a80-84ad-11ea-8457-eac5871c2dee.png)

pmiossec commented:
@gerhardol 

>This PR eliminates the possibility to reset current branch

No. If you select the first line which is in bold "Current branch: {name of the current branch}", you've got the exact same behavior as before...

> ( I considered pushing to your branch).

You could. As @mstv already did it. This PR is intended for contributions as I can't have the time to fix all that should be.

@mstv & @gerhardol 

>, I want to have two menu items

I don't have a strong feeling about that. The initial development was not like that.
I think it's also slightly better for a good user understanding of what he is doing.  So let's go with that.

I like the quick branch selection in a submenu (that's why I kept it reducing the list to the 15 or 20 more recently updated). If that's still a performance problem, probably we could remove it. Not sure if I will like it but it should not be a big deal as it is not a feature intended to be used very very frequently...

gerhardol commented:
> > This PR eliminates the possibility to reset current branch

> No. If you select the first line which is in bold "Current branch: {name of the current branch}", you've got the exact same behavior as before...

Pushed an update to change
![image](https://user-images.githubusercontent.com/6248932/79993408-053c3400-84b5-11ea-8a05-c006366d9662.png)


> > , I want to have two menu items

> I like the quick branch selection in a submenu (that's why I kept it reducing the list to the 15 or 20 more recently updated). If that's still a performance problem, probably we could remove it. Not sure if I will like it but it should not be a big deal as it is not a feature intended to be used very very frequently...

Yes, this is good, this is not a performance problem

Maybe a selection with there items is better, where Recent branch are listed

But the current implementation may be OK, will digest

mstv commented:
Now with two separate menu items and much less code.
But with the automatically dropped down branch list, it should allow equally quick operation.

![grafik](https://user-images.githubusercontent.com/36601201/80043891-35122880-8503-11ea-8814-4d11e828d0ec.png)
![grafik](https://user-images.githubusercontent.com/36601201/80044074-b669bb00-8503-11ea-8161-b7c27a551f6d.png)

gerhardol commented on GitUI/HelperDialogs/FormResetBranch.cs:
```diff
@@ -0,0 +1,71 @@
+
+            if (Branches.Text.Length == 0)
+            {
+                Branches.DroppedDown = true;
```
I prefer to not pull down the branches by default, it seems like unintentional to get the list in your face.
Resetting any branch is not done very often, it should not be a problem to click once more

mstv replied:
> I prefer to not pull down the branches by default.

A branch must be selected. What do the GUI designers say?

pmiossec replied:
>What do the GUI designers say?

I'm absolutely not a GUI designers but 😃 ...

* the display when the dropdown is opening is a little strange. The list is displayed before the rest of the form...
* all the form could be totally hidden by the branch list so, on the first use, the user could have difficulties to understand how the form works
* but once it is used one time, I find it mostly good.

So, even if it is not a full win, that's interesting.
But once again, not a strong point of view for either of the solution chosen...

gerhardol replied:
I would keep the box empty and let the user press the arrow to see the branches.
No change now, I keep my comment

pmiossec commented:
> I can push the changes, but let us agree first

I don't have a strong point of view on that (except that's a useful feature that I would like to see in GE because I needed it in the past 😉 ).

So I'm Ok and  will let @mstv share its point of view.

mstv commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.Designer.cs:
```diff
@@ -228,6 +230,14 @@ private void InitializeComponent()
+            this.resetAnotherBranchToHereToolStripMenuItem.Image = global::GitUI.Properties.Images.ResetCurrentBranchToHere;
+            this.resetAnotherBranchToHereToolStripMenuItem.Name = "resetAnotherBranchToHereToolStripMenuItem";
+            this.resetAnotherBranchToHereToolStripMenuItem.Size = new System.Drawing.Size(223, 22);
+            this.resetAnotherBranchToHereToolStripMenuItem.Text = "Reset another branch to here...";
```
> I prefer "Reset any branch to here..."

no objections

gerhardol commented:
Rebased and squashed, with the name change I wanted but without the change to dropdown
I feel this is ready to merge (with the 'dropdown' the discussion maybe open) and would like to include this if we update the translations for 3.4

pmiossec commented:
I think we must remove the currently checked out branch from the list because that's not allowed and will always fail (and must be done with the normal reset feature)

mstv commented:
@RussKie, what do you think about the automatical drop down?

RussKie commented:
> Now with two separate menu items and much less code.
> But with the automatically dropped down branch list, it should allow equally quick operation.
> 
> ![grafik](https://user-images.githubusercontent.com/36601201/80043891-35122880-8503-11ea-8814-4d11e828d0ec.png)
> ![grafik](https://user-images.githubusercontent.com/36601201/80044074-b669bb00-8503-11ea-8161-b7c27a551f6d.png)

This looks interesting.

pmiossec commented:
>I think we must remove the currently checked out branch from the list because that's not allowed and will always fail (and must be done with the normal reset feature)

I did it. Have a look.

gerhardol commented:
This PR uses ```git push ./```, ```git branch --force``` could have been used too.
The advantage with "push" should be that you get a warning if commits are lost (you have to use --force then), otherwise the behavior.
I expected that it would be possible to move the current branch, but it is not poaaible why the latest blocking of current branch is OK. Maybe my change from Any->Another should be reverted.

However, at work I got the error message "unknown remote ./", will investigate later.

mstv commented:
> However, at work I got the error message "unknown remote ./", will investigate later.

Exactly the same for me: https://github.com/gitextensions/gitextensions/issues/2876#issuecomment-616775518 Join the club! :(
So we could switch to the alternate git command you proposed.
I think it will not be too dangerous: One must select a branch and click OK. ... and there is the reflog.

pmiossec commented:
Another thing to do: don't close the "reset form" if reset failed. It's boring to have to reopen and select the branch just because you forgot to check "Force reset".

vbjay commented:
So frankly I think pushing from this type of functionality is foolish. I
think all resets should be done locally and then a push can be done with
force with lease if needed later. I'd much rather see the changed branch
info and then push second instead of all at once.

On Tue, May 5, 2020, 8:21 PM Philippe Miossec <notifications@github.com>
wrote:

> Another thing to do: don't close the "reset form" if reset failed. It's
> boring to have to reopen and select the branch just because you forgot to
> check "Force reset".
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8026#issuecomment-624375244>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMMCLMCFBXPURY6QKODRQCUSNANCNFSM4MNSTWJQ>
> .
>

mstv commented:
This is not about remote branches, Jay. `Force` has a different meaning here. See the explanation label in the screenshot of the dialog above.

Can you exchange a branch of a GH PR? I guess this is impossible.

vbjay commented:
> This PR uses ```git push ./```, ```git branch --force``` could have been used too.
> The advantage with "push" should be that you get a warning if commits are lost (you have to use --force then), otherwise the behavior.
> I expected that it would be possible to move the current branch, but it is not poaaible why the latest blocking of current branch is OK. Maybe my change from Any->Another should be reverted.
> 
> However, at work I got the error message "unknown remote ./", will investigate later.
@mstv   it was this that made me want to make sure we don't do remote branches.

pmiossec commented:
>Another thing to do: don't close the "reset form" if reset failed.

done.

gerhardol commented:
> Exactly the same for me: #2876 (comment) Join the club! :(

using ./ is not mentioned in the documentation and I find no relevant search hits for this specific situation.
https://git-scm.com/docs/git-push#_git_urls
./ can be prefixing files for sure, but for local repos file:// is mentioned

The following fails for me (from mstv example) - same Git (2.26.2)
```git.exe push ./ "303f3f4023cdbf72f0c3290f603871e320cd0ed8":"refs/heads/x" -f```

The following is OK:
```git.exe push  file:///path/to/repo "303f3f4023cdbf72f0c3290f603871e320cd0ed8":"refs/heads/x"```
(```--force``` is optional)
path can be both file:///c/dir/repo or file://c:/dir/repo
or
```git.exe branch --force x "303f3f4023cdbf72f0c3290f603871e320cd0ed8"```

Unless someone beats me, i will update with the first variant and add some comments in the code (push is slightly safer)

gerhardol commented:
Updated, also ArgumentBuilder and revert to Reset Another Branch, rebased

When reviewed, squash and merge

pmiossec commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -227,6 +227,25 @@ public static ArgumentString ResetCmd(ResetMode mode, string commit = null, stri
+        {
+            return new GitArgumentBuilder("push")
+            {
+                string.Format("file://{0}", repoPath),
```
in this case, string interpolation is more readable...

gerhardol replied:
this and next line can be combined?

mstv replied:
I think Philippe mainly means:
```suggestion
                $"file://{repoPath}",
                $"{targetId}:{gitRef}",
```

mstv commented on GitCommands/Git/GitCommandHelpers.cs:
```diff
@@ -227,6 +227,25 @@ public static ArgumentString ResetCmd(ResetMode mode, string commit = null, stri
 
+        /// <summary>
+        /// Push a local reference to a new commit
+        /// This is similar to "git branch --force "branch" "commit", except that uo get a warning if commits are lost
```
final nit:
```suggestion
        /// This is similar to "git branch --force "branch" "commit", except that you get a warning if commits are lost
```

pmiossec commented:
@gerhardol 

>However, at work I got the error message "unknown remote ./", will investigate later.

I got it also (I suspect that a `.` in the name of the branch causes that).
Your change fixes it, so 👍

gerhardol commented:
@msftbot merge in 14 days

gerhardol commented:
Squashed
I believe this is ready and should be merged - for me it could be included in 3.4, but it affects translations
I setup a merge in two weeks to give some time for @RussKie to adjust

pmiossec commented:
I have made some more work on this PR...
Please, do a last review. Especially the error message to see if it's good english.

I have:
* done missing translation
* renamed form (sorry for the review)
* cleaned and simplified code in `FormResetAnotherBranch`
* Move ref filtering that was done partly in 2 places (before form opening and when form opening) in only one place.

mstv commented:
The AppVeyor build is broken:
```
Downloading vswhere
task.logissue type=error(InitializeToolset) The remote server returned an error: (403) Forbidden.
 ```
All `.binlog`s but `localise.binlog` are missing.

mstv commented on GitUI/HelperDialogs/FormResetAnotherBranch.cs:
```diff
@@ -0,0 +1,85 @@
+
+        private void FormResetAnotherBranch_Load(object sender, EventArgs e)
+        {
+            if (_revision == null)
+            {
+                throw new Exception(Strings.NoRevision);
+            }
```
I'd rather do this check in the ctor and use the slightly more understandable `OperationNotSupportedException`.

mstv commented on GitUI/HelperDialogs/FormResetCurrentBranch.cs:
```diff
@@ -67,7 +67,7 @@ private void FormResetCurrentBranch_Load(object sender, EventArgs e)
             if (Revision == null)
             {
-                throw new Exception("No revision");
+                throw new Exception(Strings.NoRevision);
```
ditto

pmiossec commented:
Done. Squashed. Rebased.

mstv commented:
Run `build.cmd -loc` again, please.

pmiossec commented:
>Run build.cmd -loc again, please.

Yes, I saw and I did it but the result surprised me. I'm trying to understand...

gerhardol commented:
> I have made some more work on this PR...

Then I made a change, to move OK, Cancel to Strings

> Yes, I saw and I did it but the result surprised me. I'm trying to understand...

Something is strange with the translation app...
When trying this a couple of days ago, the moved strings were not removed from English.xlf, when trying earlier today all in ResetAnother was removed but now the result was as expected.

pmiossec commented:
>Something is strange with the translation app...

For my case, the behavior was normal.
That's the change asked by @mstv  to check the `revison` parameter and throw that prevent the translation of the form and so removed all the string from the translation file.

The problem didn't occurred for `FormResetCurrentBranch` so I was finding that strange. I have adopted the same strategy that is to add a private parameterless constructor fot the translation App and the Designer.

RussKie commented on GitUI/HelperDialogs/FormResetCurrentBranch.cs:
```diff
@@ -31,7 +31,7 @@ private FormResetCurrentBranch()
             : base(commands)
         {
-            Revision = revision;
+            Revision = revision ?? throw new NotSupportedException(Strings.NoRevision);
```
Throwing in the ctor prevents the object from being disposed

gerhardol replied:
We could change back to throwing in the load again, but you get the NBug anyway?

mstv replied:
The class must not be used without a revision.

pmiossec replied:
Throwing in the load method seems a good compromise. Is everyone OK?

gerhardol replied:
> Throwing in the load method seems a good compromise. Is everyone OK?

Yes, change back to that, it is slightly better as a template and if you dismiss the NBug
Whoever doing it, squash and rebase as well so it is ready for 3.4

mstv replied:
I won't block this. Though you cannot convince me that this was a clear approach. It is not worth having a further discussion on the philosophy of C# regarding this.
But the `null` check could be done by a factory method, which then calls the private ctor.

mstv replied:
done -- not squashed and rebased yet for easier review

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -1493,6 +1493,17 @@ private void ResetCurrentBranchToHereToolStripMenuItemClick(object sender, Event
+                return;
+            }
+
+            var frm = new FormResetAnotherBranch(UICommands, LatestSelectedRevision);
+            frm.ShowDialog(this);
```
```suggestion
            using (var frm = new FormResetAnotherBranch(UICommands, LatestSelectedRevision))
            {
                frm.ShowDialog(this);
            }
```

gerhardol replied:
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations
```
            using var frm = new FormResetAnotherBranch(UICommands, LatestSelectedRevision);
            frm.ShowDialog(this);
```

RussKie commented:
The warning label overlaps and partially covers buttons (DPI 144dpi (150% scaling))
![image](https://user-images.githubusercontent.com/4403806/82207976-0e9eac00-994e-11ea-872c-8021017c9417.png)


I think it would be more appropriate to place the label above all other elements. Being the last element it feels like something is "_btw fyi_" rather than "_IMPORTANT!_". The warning label wrapping looks awkward, but don't have a better suggestion at this stage.
The icon should be "warning" rather than "info", and it needs to be scaled to be of a greater size.

RussKie commented on GitUI/HelperDialogs/FormResetCurrentBranch.cs:
```diff
@@ -22,13 +22,10 @@ public enum ResetType
             Hard
         }
 
-        [Obsolete("For VS designer and translation test only. Do not remove.")]
-        private FormResetCurrentBranch()
-        {
-            InitializeComponent();
-        }
```
Without a default ctor it is impossible to open the form in the designer

mstv replied:
Sorry, this was for those who can read. Restored.

mstv commented:
`git reset && git checkout` modifies the `English.xlf` if it was changed both in the master and in the PR branch. But it must not be changed by the tweaking of the commit hash.
That's why I restored the approach using `git commit`, which seems to work.

Remains Igor's request to move the warning message. Form layout is none of my strength, especially not using the VS designer. Volunteers?

pmiossec commented:
An attempt to improve UI:
![image](https://user-images.githubusercontent.com/460196/82269837-2b99a600-9973-11ea-8a73-5ac52e17eee5.png)

1. Update form design to minimize risk of label overwrite

* Increase form width (and so label and commit description box width)
* Put warning label before checkbox
* Remove label autosize (no more overlapping possible)
* increase form width to prevent wrapping in most of the case
* indent checkbox to better express dependency with warning text above

2. More explicit checkbox text "Force reset for a non-fast-forward reset"

3. Non luck in trying to resize picturebox :(

Point of view?

RussKie commented:
> * Increase form width (and so label and commit description box width)
> * Remove label autosize (no more overlapping possible)

This may not work well with translations, so the label needs to be fluid, auto scale/layout.
I'll get some time and send you my proposal.

> \2. More explicit checkbox text "Force reset for a non-fast-forward reset"

Nice!

RussKie commented:
> Form layout is none of my strength, especially not using the VS designer. Volunteers?

I'll give a hand.

pmiossec commented:
>This may not work well with translations, so the label needs to be fluid, auto scale/layout.

@RussKie another try before you put an hand on that.
Warning label set again at auto resize.
**But put also the form as auto resize** to keep the label on 2 lines (much more readable) and prevent vertical overlapping.

Example of the worst case:
![image](https://user-images.githubusercontent.com/460196/82325979-6b48a800-99dc-11ea-9b3f-9ffdf04c9ed2.png)

If that's not good, I let you handle that ;)

pmiossec commented:
>Squash and rebase on master for another try?

I have squashed all commits except one (by @mstv ) related to the localization in the AppVeyor build.
I don't really get the goal of this commit and I don't know if we can/must remove it from this PR.
Or if this should end up in another PR...

mstv commented:
> The appveyor commit should be removed

Moved to another PR!

mstv commented:
This PR did not pass the translation check after changes to the `English.xlf` on `master`.

RussKie commented:
Just to let you know I'm reviewing the changes.

pmiossec commented:
>This PR did not pass the translation check after changes to the English.xlf on master.

It seems that the build passed. Is there something that I don't understand?

RussKie commented:
![image](https://user-images.githubusercontent.com/4403806/82562103-f4680680-9bb7-11ea-86c1-bc12bd8c2b1e.png)

mstv commented:
Since the rebase of this branch, of course there are no concurrent changes to `English.xlf` anymore. Hence the build succeeds now without #8132.

RussKie commented:
Please try my changes. Few commits for the ease of reviewing, squash if/when happy.

RussKie commented:
Aye, good point.
The less elements the better.

pmiossec commented:
Closing in favor of an upcoming PR to merge in `3.4` branch instead...

gerhardol commented:
You can just edit this PR...

pmiossec commented:
> You can just edit this PR...

I think we can't change the branch to merge to...

gerhardol commented:
For open PRs you can

pmiossec commented:
Ok. I never found how. I will try to find for the next time (as I can't reopen this PR because it has been force pushed. GitHub don't let me.)

See you in #8140

mstv commented:
I have not noticed this was possible. But the `Edit` button beside the title really does.
